types/vega/encode.d.ts

```diff
@@ -0,0 +1,180 @@
+    strokeWidth?: ProductionRule<NumericValueRef>;
+    strokeOpacity?: ProductionRule<NumericValueRef>;
+    strokeDash?: ProductionRule<ScaledValueRef<number[]>>;
+    strokeDashOffset?: ProductionRule<NumericValueRef>;
+    zindex?: ProductionRule<NumericValueRef>;
+    cursor?: ProductionRule<StringValueRef>;
+    tooltip?: ProductionRule<StringValueRef>;
+    xfocus?: ProductionRule<NumericValueRef>;
+    yfocus?: ProductionRule<NumericValueRef>;
+}
+export declare type Align = 'left' | 'center' | 'right';
+export interface AlignProperty {
+    align?: ProductionRule<ScaledValueRef<Align>>;
+}
+export interface DefinedProperty {
+    defined?: ProductionRule<BooleanValueRef>;
+}
+export interface ThetaProperty {
+    theta?: ProductionRule<NumericValueRef>;
+}
+export interface ArcEncodeEntry extends EncodeEntry {
+    startAngle?: ProductionRule<NumericValueRef>;
+    endAngle?: ProductionRule<NumericValueRef>;
+    padAngle?: ProductionRule<NumericValueRef>;
+    innerRadius?: ProductionRule<NumericValueRef>;
+    outerRadius?: ProductionRule<NumericValueRef>;
+    cornerRadius?: ProductionRule<NumericValueRef>;
+}
+export declare type Orientation = 'horizontal' | 'vertical';
+export interface AreaEncodeEntry extends LineEncodeEntry {
+    orient?: ProductionRule<ScaledValueRef<Orientation>>;
+}
+export interface GroupEncodeEntry extends RectEncodeEntry {
+    clip?: ProductionRule<BooleanValueRef>;
+}
+export declare type Baseline = 'top' | 'middle' | 'bottom';
+export interface ImageEncodeEntry extends EncodeEntry, AlignProperty {
+    url: ProductionRule<StringValueRef>;
+    aspect?: ProductionRule<BooleanValueRef>;
+    baseline?: ProductionRule<ScaledValueRef<Baseline>>;
+}
+export declare type Interpolate = 'basis' | 'bundle' | 'cardinal' | 'catmull-rom' | 'linear' | 'monotone' | 'natural' | 'step' | 'step-after' | 'step-before';
+export interface LineEncodeEntry extends EncodeEntry, DefinedProperty {
+    interpolate?: ProductionRule<ScaledValueRef<Interpolate>>;
+    tension?: ProductionRule<NumericValueRef>;
+}
+export interface PathEncodeEntry extends EncodeEntry {
+    path?: ProductionRule<StringValueRef>;
+}
+export interface RectEncodeEntry extends EncodeEntry {
+    cornerRadius?: ProductionRule<NumericValueRef>;
+}
+export declare type RuleEncodeEntry = EncodeEntry;
+export interface ShapeEncodeEntry extends EncodeEntry {
+    shape?: ProductionRule<StringValueRef>;
+}
+export declare type SymbolShapes = 'circle' | 'square' | 'cross' | 'diamond' | 'triangle-up' | 'triangle-down' | 'triangle-right' | 'triangle-left' | string;
+export interface SymbolEncodeEntry extends EncodeEntry {
+    size?: ProductionRule<NumericValueRef>;
+    shape?: string | ProductionRule<ScaledValueRef<SymbolShapes>>;
+}
+export declare type TextBaseline = 'alphabetic' | Baseline;
+export declare type TextDirection = 'ltr' | 'rtl';
+export declare type FontWeight =
+    | 'normal'
+    | 'bold'
+    | '100'
```

Comment:
I think these can be numbers, right?

Reply:
Not according to https://vega.github.io/schema/vega/v3.json:

```
        "fontWeight": {
          "anyOf": [
            {
              "type": "string"
            },
            {
              "$ref": "#/refs/signal"
            },
            {
              "$ref": "#/refs/expr"
            },
            {
              "$ref": "#/refs/paramField"
            }
          ],
          "default": "normal"
        },
```

Reply:
If you change the `fontWeight` in [this example](https://vega.github.io/editor/#/examples/vega/crossfilter-flights) to a number it will show a bunch of warnings due to the schema violation, though it still seems to work. We could allow arbitrary numbers, but it's less type safe because then you can specify invalid weights like -3.141592653.

Reply:
I meant `type Foo = 'bold' | 100 | 200 | 300;` instead of `type Foo = 'bold' | '100' | '200' | '300';`

Reply:
> Not according to https://vega.github.io/schema/vega/v3.json:

That's odd.

Reply:
Adding `100 | 200 | 300 | ...` makes sense to me.

Reply:
I just tested it and numbers work. I will file an issue. 

Yet another reason to generate the JSON schema ;-)

Reply:
https://github.com/vega/vega-parser/issues/70